src/library.coffee

```diff
@@ -122,21 +137,29 @@ class Library extends EventEmitter
     return null
 
   _updateComponents: (components) ->
-    names = Object.keys components
+    names = []
     for name, comp of components
       if not comp
         # removed
         @components[name] = null
+        names.push name if names.indexOf(name) is -1
         continue
+      discovered = cleanComponentDefinition comp
       existing = @getComponent name
-      unless existing
+      unless existing?.definition
         # added
-        @components[name] = comp
+        @components[name] = discovered
+        names.push name if names.indexOf(name) is -1
+        continue
+      unless JSON.stringify(existing.definition) is JSON.stringify(discovered.definition)
+        # updated
+        for k, v of discovered
+          @components[name][k] = v
+        names.push name if names.indexOf(name) is -1
         continue
-      # update
-      for k, v of comp
-        existing[k] = v
 
+    # Skip sending components-changed if nothing changed
+    return unless names.length
```

Comment:
```
if names.length
  @emit 'components-changed', names, @components
```
is much clearer than a conditional return (only really acceptable for failed preconditions, imo)